hip/components/cooperative_groups.hip.hpp

```diff
@@ -224,30 +249,54 @@ class thread_block_tile {
 
     __device__ __forceinline__ int any(int predicate) const noexcept
     {
-        static_assert(Size == kernels::hip::config::warp_size,
-                      "Hip does not have subwarp any.");
-        return __any(predicate);
+#if GINKGO_HIP_PLATFORM_HCC
+        if (Size == config::warp_size) {
+            return __any(predicate);
+        } else {
+            return (__ballot(predicate) & data_.mask) != 0;
+        }
+#else
+        return __any_sync(data_.mask, predicate);
+#endif
     }
 
     __device__ __forceinline__ int all(int predicate) const noexcept
     {
-        static_assert(Size == kernels::hip::config::warp_size,
-                      "Hip does not have subwarp all.");
-        return __all(predicate);
+#if GINKGO_HIP_PLATFORM_HCC
+        if (Size == config::warp_size) {
+            return __all(predicate);
+        } else {
+            return (__ballot(predicate) & data_.mask) == data_.mask;
+        }
+#else
+        return __all_sync(data_.mask, predicate);
+#endif
     }
 
     __device__ __forceinline__ config::lane_mask_type ballot(
         int predicate) const noexcept
     {
-        static_assert(Size == kernels::hip::config::warp_size,
-                      "Hip does not have subwarp ballot.");
-        return __ballot(predicate);
+#if GINKGO_HIP_PLATFORM_HCC
+        if (Size == config::warp_size) {
+            return __ballot(predicate);
+        } else {
+            return (__ballot(predicate) & data_.mask) >> data_.lane_offset;
+        }
+#else
+        if (Size == config::warp_size) {
+            return __ballot_sync(data_.mask, predicate);
+        } else {
+            return __ballot_sync(data_.mask, predicate) >> data_.lane_offset;
```

Comment:
If the lanemask is 11110000 and all predicates are 1, the result will be 1111 not 11110000, is correct?
I only found the similar implementation `_packLane` in `cooperative_groups.h` of cuda, but not in official documentation.

Reply:
Yes, that is correct. The documentation is not very specific in that regard, but as I understand it, the bit indices correspond to the `group.thread_rank()` of a thread, not its warp lane. Otherwise, thread block tiles would not be very portable for sub-warp sizes. The implementation whose behavior I wanted to mirror is actually in `__thread_block_tile_base`, not `coalescing_group`